First draft of Capture statistics for SCALAR functions/UDF. (#1)
ScrapCodes/rfcs: RFC-0003-functions-stats.md

# **RFC0003 for Presto**

## Capture statistics for SCALAR functions/UDF

Proposers

* Anant Aneja (@aaneja)
* Prashant Sharma(@ScrapCodes)

## [Related Issues]

1. https://github.com/prestodb/presto/issues/20513
2. https://github.com/prestodb/presto/issues/19356

Related issues may include GitHub issues, PRs or other RFCs.

## Summary

Provide an easy and backwards compatible way for the Presto SCALAR functions(UDFs) to configure a callback or provide certain params via annotation
to compute how a function impacts source statistics (.e.g distinct value count, null fraction etc...). SCALAR functions have certain 

> characteristics (e.g. they do not impact output row counts) which can be used to provide more options for stats calculation than just

Reviewer: It might be beneficial to introduce `VariableStatisticsEstimate` and its fields.

Author: Added a section in `Background`, do you intend we should have it in summary. Wanted to keep summary section short.

registering callback.

> A SCALAR function `upper(COLUMN1)` can be optimized if it could just say `propagateStatistics: true` i.e. just reuse source statistics of `COLUMN1`.

Reviewer: ```suggestion
A SCALAR function `upper(COLUMN1)` can be optimized if we can reuse source statistics of `COLUMN1`.
```

> Another example is `is_null(COLUMN1)` can be optimized if it could say `distinctValueCount: 2` and `nullFraction: 0.0` etc...

Reviewer: ```suggestion
Another example is `is_null(COLUMN1)` can be optimized if it could say `distinctValueCount: 2` and `nullFraction: 0.0`.
```

> We can add a special case for each example as above or come up with something more general, the current codebase adds special case for functions `CAST(...)`,

Reviewer: Can you link a reference to the relevant location(s) in the codebase and describe how it is accomplished?

> `is_null(Slice)`, and all arithmetic operators. This proposal discusses alternatives to adding special cases or Heuristics for optimising these udf or functions.

Reviewer 2: nit: lower case `heuristics`.
nit : reword to `optimizing scalar UDFs`

If none of the parameters can fit a certain functions characteristic then it can use a user provided support function(or a callback) for determining
function statistics.

> Finally, exact statistics cannot be known in advance e.g. the distinctValueCount for column will be different after applying `upper(Col1)` function,

Reviewer: ```suggestion
Finally, exact statistics cannot be known in advance e.g. the distinct value count for a column will be different after applying `upper(Col1)` function,
```

and it depends on the data. A closer estimate can be arrived at - by either using a stochastic approach or sampling.

## Background

> A function is a black box for a SQL optimizer and thus hard to optimize. Issues [1] and [2] are examples of queries we could have optimized better

Reviewer 2: Let's be precise here - `The Presto optimizer cannot estimate statistics for UDFs since it is unaware of how source stats need to be transformed`

> A function is a black box for a SQL optimizer and thus hard to optimize. Issues [1] and [2] are examples of queries we could have optimized better

Reviewer: [1] and [2] don't seem to be linked properly

> if there was some way to compute source stats (e.g. ndv, null_fraction etc...) for _functions_ or _udf_. The current implementation simply applies 

Reviewer 2: `if there was a way to derive stats from source stats`

> if there was some way to compute source stats (e.g. ndv, null_fraction etc...) for _functions_ or _udf_. The current implementation simply applies 

Reviewer: nit: conciseness
```suggestion
if there was some way to compute source stats. The current implementation simply applies 
```

                             Estimates: {source: CostBasedSourceInfo, rows: 100000 (488.28kB), cpu: 3581866.00, memory: 0.00, network: 2240933.00}                                                      >
                         - ScanProject[table = TableHandle {connectorId='local_hms', connectorHandle='HiveTableHandle{schemaName=tpcds_sf1_parquet, tableName=customer, analyzePartitionValues=Optional.>
                                 Estimates: {source: CostBasedSourceInfo, rows: 100000 (488.28kB), cpu: 1340933.00, memory: 0.00, network: 0.00}/{source: CostBasedSourceInfo, rows: 100000 (488.28kB), >
                                 $hashvalue_8 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(c_birth_country), BIGINT'0')) (1:23)                                                              >
                                 LAYOUT: tpcds_sf1_parquet.customer{}                                                                                                                                   >
                                 c_birth_country := c_birth_country:varchar(20):14:REGULAR (1:23)      
```

> The functions like `CAST(COL)` and all arithmetic functions are handled in a specialised way their statistics are computed as their characteristics

Reviewer 2: Can you show these in action with some examples ?

> The functions like `CAST(COL)` and all arithmetic functions are handled in a specialised way their statistics are computed as their characteristics

Reviewer: similar comment. A reference to existing implementation would be helpful

are known in advance. Whereas a `function` in general can not optimize in the same way, because it's characteristics are not known. This proposal
is aimed at bridging this gap.

> Reference [1. Monsoon](#references). discusses various stochastic approaches to solve the same problem, a user provided stats calculator will complement such approaches 

Reviewer 2: This seems a bit out of place. Can we move this to 'Other Approaches Considered' ?

  /// `nullFraction`.
  double nullFractionAdjustFactor{1.0};

  /// An `avgRowSize`: does this function impacts the size of each row e.g. a
  /// function like md5 may produce a constant row size.
  double avgRowSize{NAN};

  /// An `avgRowSizeAdjustFactor`: does this function impacts the size of each
  /// row e.g. a function may alter the column type to a wider type
  /// e.g.varchar(20)->varchar(40).
  double avgRowSizeAdjustFactor{1.0};
};
```

Add a virtual function to `VectorFunctions.h` which a function can override with the function specific 
statistics propagation override, e.g. 

```cpp
@@ -109,6 +160,11 @@ class VectorFunction {
     return false;
   }
 
+  // How this function impacts statistics.
+  virtual std::optional<VectorFunctionStatistics> statsResolver() const {
+    return std::nullopt;
+  }
+
```

```cpp
@@ -129,6 +129,10 @@ class UpperLowerTemplateFunction : public exec::VectorFunction {
   bool propagateStringEncodingFromAllInputs() const override {
     return true;
   }
+
+  std::optional<exec::VectorFunctionStatistics> statsResolver() const override {
+    return std::optional{exec::VectorFunctionStatistics{true}};
+  }
 };
```

>  * JAVA: Provide a new Annotation i.e. `ScalarFunctionStatsCalculator` in java with fields as follows:

Reviewer: In CPP it looks like you're providing a function. This would potentially allow for more complex implementations in the future such as the sampling approach described earlier. With just an annotation on the scalar function class on the Java side, I don't believe that would be possible.

Do you think it would be possible to annotate a java function within the scalar function definition to allow custom logic other than `propagate=true`? Or provide a custom calculator implementation?

Author: Hi @ZacBlanco , thanks for the feedback and I am still working on it.

In CPP even though it seems to be a function, but it is not a very powerful abstraction. The abstraction does not allow a way to capture the stats of the input columns and transform them. It is slightly better than JAVA's annotation.

> Do you think it would be possible to annotate a java function within the scalar function definition to allow custom logic other than propagate=true? Or provide a custom calculator implementation?

Yes, I have a plan for it i.e. Phase 3 of implementation.

 * applied to input columns on which this function operates on. 1.0 is the default value indicating
 * Propagate input columns stats as is.
 * </li>
 * <li>
 * 2. How to estimate number of distinct values. Does this function produces an intersection of
 * Distinct value counts of each `input column param` or a union.
 * </li>
 * </ul>
 */
@Retention(RUNTIME)
@Target({METHOD, TYPE})
public @interface ScalarFunctionStatsCalculator
{
  /**
   * Disable or enable propagate stats behaviour. A function (e.g. upper(Slice)) may not alter any source stats for the
   * input column passed as an argument. In this case it might be possible to just set propagate stats as true.
   * In cases where number of input columns are more than one, one has few choices i.e. is the resulting statistics
   * a union of all columns, Max, Sum or an intersection. To achieve this behaviour please set `statsResolver` field
   * in conjunction of setting this. For more precise control consider supplying a callback via ... see Phase 3.
   */
  boolean propagateStats() default false;

  /**
   * Possible values: Max, Sum, union and intersect.
   */

>   String statsResolver() default "Max";

Reviewer 2: Will the `statsResolver` apply the same mechanism on all types of stats ? E.g say I want to choose the min of the distinct value but max of the null fraction, will that be possible ?

Reviewer 2: Can you illustrate, with an example, a multi parameter function where it would make sense to use one of these stats resolver strategies

Author: Great point, 
Can we switch to:
`distinctValueCountMapper : "x + y" or "max(x, y)"`
This is much more powerful, the value x and y represents parameters' `x` and `y`'s distinct value count, and expression can be solved using : [grammer](https://github.com/prestodb/presto/blob/master/presto-parser/src/main/antlr4/com/facebook/presto/type/TypeCalculation.g4).

e.g.

```java
    @Description("returns index of first occurrence of a substring (or 0 if not found)")
    @ScalarFunction("strpos")
    @LiteralParameters({"x", "y"})
    @SqlType(StandardTypes.BIGINT)
    @ScalarFunctionStatsCalculator(propagateStats = true, avgRowSize = 64, distinctValueCountMapper = "x*y" )
     // There may be repeated values at different Rows, but their orders will be diff
    // as a result ndv will increase geometrically.
    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
    {
        return stringPositionFromStart(string, substring, 1);
    }
```

Reviewer: I feel like introducing more annotations/string expressions to parse requires additional complexity that provides little benefit. You could just introduce a java function and write code to compute it directly. It would make compiling the logic for compiling the function metadata quite simple as you have a 1:1 mapping of function-stats calculations

You could implement an annotation which is shorthand for basic propagations (e.g. `@ScalarFunctionStats(propagate = true)` to make it a little easier for authors. My opinion though is that if we're requiring the ability interpret expressions you may as well allow the author to write the code to do it directly.

```java
    @Description("returns index of first occurrence of a substring (or 0 if not found)")
    @ScalarFunction("strpos")
    @SqlType(StandardTypes.BIGINT)
    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
    {
        return stringPositionFromStart(string, substring, 1);
    }

    @ScalarFunctionStatsCalculator
    public static VariableStatisticsEstimate statsPropgation(VariableStatisticsEstimate inputs...)
    {
        // inherits values from arg at position 0 by default
        return VariableStatsEstimate.builder(inputs[0])
                       .setAverageRowSize(64)
                       .setDistinctValuesCount(inputs[0].getDistinctValues() * inputs[1].getDistinctValues())
                       .build();
    }   
```
